chalice/chalicelib/request_handler.py

```diff
@@ -0,0 +1,128 @@
+
+        try:
+            content = s3_blob_store.get(bucket=REQUEST_STATUS_BUCKET_NAME, key=key)
+            value = json.loads(content)[fieldname]
+            return value
+        except Exception as e:
+            raise e
+
+    @staticmethod
+    def get_request_body(request_id: str) -> dict:
+        """
+        Get request body from a request status file in s3.
+        :param request_id: Matrices concatenation request id.
+        :return: Request body.
+        """
+        key = request_id + JSON_SUFFIX
+
+        try:
+            content = s3_blob_store.get(bucket=REQUEST_STATUS_BUCKET_NAME, key=key)
+            body = json.loads(content)
+            return body
+        except Exception as e:
+            raise e
+
+    @staticmethod
+    def update_request(
+            bundle_uuids: List[str],
+            request_id: str,
+            job_id: str,
+            status: RequestStatus,
+            time_spent_to_complete="",
+            reason_to_abort="") -> None:
+        """
+        Update the request status json file in s3 bucket if exists. Otherwise,
+        create a new status file.
+
+        :param bundle_uuids: A list of bundle uuids.
+        :param request_id: Matrices concatenation request id.
+        :param job_id: Matrices concatenation job id.
+        :param status: Request status to update.
+        :param time_spent_to_complete: Time spent to complete the concatenation job.
+        :param reason_to_abort: Request abort reason.
+        """
+        request_key = request_id + JSON_SUFFIX
+
+        # Check whether request status file exists in the s3 bucket; Create one if not.
+        try:
+            RequestHandler.get_request_body(request_id=request_id)
+        except BlobNotFoundError:
+            with tempfile.TemporaryFile(dir=TEMP_DIR, suffix=JSON_SUFFIX) as f:
+                s3_blob_store.upload_file_handle(
+                    bucket=REQUEST_STATUS_BUCKET_NAME,
+                    key=request_key,
+                    src_file_handle=f
+                )
+        except Exception as e:
+            raise e
+
+        # Create a request based on a template dict
+        request = REQUEST_TEMPLATE.copy()
+        request["bundle_uuids"] = bundle_uuids
+        request["status"] = status.name
+        request["request_id"] = request_id
+        request["job_id"] = job_id
+        request["time_spent_to_complete"] = time_spent_to_complete
+        request["reason_to_abort"] = reason_to_abort
+
+        if status == RequestStatus.DONE:
+            # Key for merged matrix stored in s3 bucket
+            mtx_key = request_id + ".loom"
+
+            mtx_url = "s3://{}/{}".format(
+                MERGED_MTX_BUCKET_NAME,
+                mtx_key
+            )
+            request["merged_mtx_url"] = mtx_url
+
+        try:
+            s3_request_obj = RequestHandler.s3.Object(REQUEST_STATUS_BUCKET_NAME, request_key)
+            s3_request_obj.put(Body=json.dumps(request))
```

Comment:
I'm pretty sure S3 only offers eventual consistency for PUTs that overwrite an existing file. This will probably lead to lots of weird bugs.

Reply:
Hey Marcus. Can you please elaborate more this issue? I know that S3 offers eventual consistency model, but I do not see any potential bug that could arise here. Can you give me an example? Thanks.

Reply:
Well, I think you've already thought about one case a bit in this comment:

          2. Before receiving the msg from SQS queue, a duplicate mtx
             concatenation request has been initialized. As a result,
             the job id in the status file in s3 has been updated to
             a new value.

The problem is you don't know which version of the status file in S3 you're going to get. It could be the new one written by the duplicate matrix request, or it could be the first one with a job_id that matches what's in the queue.

Reply:
Okay. If it gets the new one written by the duplicate matrix request, it will find that there is a inconsistency between the job id it receives and the one in the S3 file. As a result, it will abandon processing a merge on it. This behavior is expected.

If it gets the old one instead of the new one. It will merge the matrices because of the match in job ids. And, when the duplicate request being polled from the queue and processed by the lambda function, since there is still a match in job ids, that requests will still be handled. As a result, we do a matrices concatenation twice.

Thus, it seems like that in the worst case, the AWS S3 consistency model will cause the current version of matrix service not being able to avoid duplicate matrices concatenation. Is my understanding correct?

Reply:
Hmm, yeah this is good to think through. Consider the follow though:

- Two nearly simultaneous POSTs to /concat get made with the same list of bundle ids
- Neither sees an object named request_id_1.json (making up a name here). So each generates a different job_id, each PUTs an object with a different job_id into S3 and a message into the queue. Call these request_id_1.json[job_id_1], request_id_1.json[job_id_2], message[job_id_1], and message[job_id_2].

Now the queue triggers two lambda executions, one for each message. 

- Suppose the first lambda gets message[job_id_1]. It then looks at S3 for request_id_1.json. We don't know if it's going to get request_id_1.json[job_id_1] or request_id_1.json[job_id_2]. But suppose it gets the latter. It's going to see that the message and S3 job_ids don't match, and it's going to halt.
- Now the second lambda gets message[job_id_2]. It also looks at S3 for request_id_1.json, and we still don't know which version it's going to get, but suppose we're unlucky and it gets request_id_1.json[job_id_1]. Just like the first lambda it also halts.

Now we have a matrix request in a permanent INITIALIZED state, but nothing is happening. I think this set of events is possible, but I might be overlooking something.

More generally, it's worthing looking through the code at every `get_request_field` or `get_request_body` and considering what could happen if you got a stale or inconsistent S3 object.

Also, I don't think this needs to block merging this PR. But if we decide we need to address it, we should merge and file an issue.

Reply:
I see your point. I will take carefully look at every operation related to `get` things from S3.

Reply:
Hey Marcus. Do you think enable S3 versioning can help to address this problem. Whenever we try to update the request status file, we always create a new version of it instead of overwriting the old one. And, when we try to GET object from S3, we always get the newest version of it. Doing so can guarantee the `read-after-write consistency` provided by S3 because there is no overwrite happens.

Reply:
I don't think so. Specifically, I don't think this is true: "And, when we try to GET object from S3, we always get the newest version of it." The S3 consistency model stays the same; you just can refer to an object that otherwise would have been overwritten. The solution is probably to switch to an AWS product with better consistency guarantees. DynamoDB might be enough.